File: android/src/main/kotlin/com/example/vonage_video_flutter/VonageVideoFlutterPlugin.kt
package com.example.vonage_video_flutter

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformViewRegistry

class VonageVideoFlutterPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private lateinit var sessionManager: VonageSessionManager
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext

        // Initialize method channel
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "vonage_video_flutter")
        methodChannel.setMethodCallHandler(this)

        // Initialize event channel
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "vonage_video_flutter/events")
        sessionManager = VonageSessionManager(context)
        eventChannel.setStreamHandler(sessionManager)

        // Register platform views
        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "vonage_video_flutter/publisher_view",
            VonagePublisherViewFactory(sessionManager, StandardMessageCodec.INSTANCE)
        )

        flutterPluginBinding.platformViewRegistry.registerViewFactory(
            "vonage_video_flutter/subscriber_view",
            VonageSubscriberViewFactory(sessionManager, StandardMessageCodec.INSTANCE)
        )
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "connect" -> {
                val apiKey = call.argument<String>("apiKey")
                val sessionId = call.argument<String>("sessionId")
                val token = call.argument<String>("token")

                if (apiKey != null && sessionId != null && token != null) {
                    sessionManager.connect(apiKey, sessionId, token)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGS", "Missing required arguments", null)
                }
            }
            "disconnect" -> {
                sessionManager.disconnect()
                result.success(null)
            }
            "sendSignal" -> {
                val type = call.argument<String>("type")
                val data = call.argument<String?>("data")
                val connectionId = call.argument<String?>("connectionId")

                if (type != null) {
                    sessionManager.sendSignal(type, data, connectionId)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGS", "Missing type argument", null)
                }
            }
            "initPublisher" -> {
                val name = call.argument<String?>("name")
                val publishAudio = call.argument<Boolean>("publishAudio") ?: true
                val publishVideo = call.argument<Boolean>("publishVideo") ?: true
                val cameraPosition = call.argument<String>("cameraPosition") ?: "front"

                sessionManager.initPublisher(name, publishAudio, publishVideo, cameraPosition)
                result.success(null)
            }
            "publish" -> {
                sessionManager.publish()
                result.success(null)
            }
            "unpublish" -> {
                sessionManager.unpublish()
                result.success(null)
            }
            "setPublishAudio" -> {
                val enabled = call.argument<Boolean>("enabled") ?: false
                sessionManager.setPublishAudio(enabled)
                result.success(null)
            }
            "setPublishVideo" -> {
                val enabled = call.argument<Boolean>("enabled") ?: false
                sessionManager.setPublishVideo(enabled)
                result.success(null)
            }
            "switchCamera" -> {
                sessionManager.switchCamera()
                result.success(null)
            }
            "subscribe" -> {
                val streamId = call.argument<String>("streamId")
                if (streamId != null) {
                    sessionManager.subscribe(streamId)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGS", "Missing streamId", null)
                }
            }
            "unsubscribe" -> {
                val streamId = call.argument<String>("streamId")
                if (streamId != null) {
                    sessionManager.unsubscribe(streamId)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGS", "Missing streamId", null)
                }
            }
            "setSubscribeToAudio" -> {
                val streamId = call.argument<String>("streamId")
                val enabled = call.argument<Boolean>("enabled") ?: false
                if (streamId != null) {
                    sessionManager.setSubscribeToAudio(streamId, enabled)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGS", "Missing streamId", null)
                }
            }
            "setSubscribeToVideo" -> {
                val streamId = call.argument<String>("streamId")
                val enabled = call.argument<Boolean>("enabled") ?: false
                if (streamId != null) {
                    sessionManager.setSubscribeToVideo(streamId, enabled)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGS", "Missing streamId", null)
                }
            }
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        sessionManager.cleanup()
    }
}

File: ios/Classes/VonageVideoFlutterPlugin.swift
import Flutter
import UIKit

public class VonageVideoFlutterPlugin: NSObject, FlutterPlugin {
    private var sessionManager: VonageSessionManager?

    public static func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: "vonage_video_flutter",
            binaryMessenger: registrar.messenger()
        )

        let eventChannel = FlutterEventChannel(
            name: "vonage_video_flutter/events",
            binaryMessenger: registrar.messenger()
        )

        let instance = VonageVideoFlutterPlugin()
        instance.sessionManager = VonageSessionManager()

        registrar.addMethodCallDelegate(instance, channel: methodChannel)
        eventChannel.setStreamHandler(instance.sessionManager)

        // Register platform views
        let publisherFactory = VonagePublisherViewFactory(sessionManager: instance.sessionManager!)
        registrar.register(
            publisherFactory,
            withId: "vonage_video_flutter/publisher_view"
        )

        let subscriberFactory = VonageSubscriberViewFactory(sessionManager: instance.sessionManager!)
        registrar.register(
            subscriberFactory,
            withId: "vonage_video_flutter/subscriber_view"
        )
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let sessionManager = sessionManager else {
            result(FlutterError(code: "NO_SESSION", message: "Session manager not initialized", details: nil))
            return
        }

        switch call.method {
        case "connect":
            guard let args = call.arguments as? [String: Any],
                  let apiKey = args["apiKey"] as? String,
                  let sessionId = args["sessionId"] as? String,
                  let token = args["token"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing required arguments", details: nil))
                return
            }
            sessionManager.connect(apiKey: apiKey, sessionId: sessionId, token: token)
            result(nil)

        case "disconnect":
            sessionManager.disconnect()
            result(nil)

        case "sendSignal":
            guard let args = call.arguments as? [String: Any],
                  let type = args["type"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing type argument", details: nil))
                return
            }
            let data = args["data"] as? String
            let connectionId = args["connectionId"] as? String
            sessionManager.sendSignal(type: type, data: data, connectionId: connectionId)
            result(nil)

        case "initPublisher":
            let args = call.arguments as? [String: Any]
            let name = args?["name"] as? String
            let publishAudio = args?["publishAudio"] as? Bool ?? true
            let publishVideo = args?["publishVideo"] as? Bool ?? true
            let cameraPosition = args?["cameraPosition"] as? String ?? "front"

            sessionManager.initPublisher(
                name: name,
                publishAudio: publishAudio,
                publishVideo: publishVideo,
                cameraPosition: cameraPosition
            )
            result(nil)

        case "publish":
            sessionManager.publish()
            result(nil)

        case "unpublish":
            sessionManager.unpublish()
            result(nil)

        case "setPublishAudio":
            guard let args = call.arguments as? [String: Any],
                  let enabled = args["enabled"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing enabled argument", details: nil))
                return
            }
            sessionManager.setPublishAudio(enabled: enabled)
            result(nil)

        case "setPublishVideo":
            guard let args = call.arguments as? [String: Any],
                  let enabled = args["enabled"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing enabled argument", details: nil))
                return
            }
            sessionManager.setPublishVideo(enabled: enabled)
            result(nil)

        case "switchCamera":
            sessionManager.switchCamera()
            result(nil)

        case "subscribe":
            guard let args = call.arguments as? [String: Any],
                  let streamId = args["streamId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing streamId", details: nil))
                return
            }
            sessionManager.subscribe(streamId: streamId)
            result(nil)

        case "unsubscribe":
            guard let args = call.arguments as? [String: Any],
                  let streamId = args["streamId"] as? String else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing streamId", details: nil))
                return
            }
            sessionManager.unsubscribe(streamId: streamId)
            result(nil)

        case "setSubscribeToAudio":
            guard let args = call.arguments as? [String: Any],
                  let streamId = args["streamId"] as? String,
                  let enabled = args["enabled"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing arguments", details: nil))
                return
            }
            sessionManager.setSubscribeToAudio(streamId: streamId, enabled: enabled)
            result(nil)

        case "setSubscribeToVideo":
            guard let args = call.arguments as? [String: Any],
                  let streamId = args["streamId"] as? String,
                  let enabled = args["enabled"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing arguments", details: nil))
                return
            }
            sessionManager.setSubscribeToVideo(streamId: streamId, enabled: enabled)
            result(nil)

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}
